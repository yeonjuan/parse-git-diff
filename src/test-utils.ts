import Context from './context';

type Validator<Args extends any[] = any[]> = (...args: Args) => boolean;

function createBooleanTester<V extends Validator>(
  validator: V,
  result: boolean
) {
  return function validTester(cases: Parameters<V>[]): void {
    cases.forEach((c) => {
      expect(validator(...c)).toBe(result);
    });
  };
}

export function createValidTester<V extends Validator>(validator: V) {
  return createBooleanTester(validator, true);
}

export function createInvalidTester<V extends Validator>(validator: V) {
  return createBooleanTester(validator, false);
}

export function createContext(diff: string, initial: number = 1) {
  const context = new Context(diff);
  for (let i = 0; i < initial - 1; i++) {
    context.nextLine();
  }
  return context;
}
