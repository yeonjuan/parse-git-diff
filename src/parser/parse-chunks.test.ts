import { createContext } from '../test-utils';
import parseChunks from './parse-chunks';

describe('parseChunks', () => {
  it('should parse chunks.', () => {
    // prettier-ignore
    const src =
`@@ -18,7 +18,7 @@ export type AnyChange = Added | Deleted | Unchanged;
 /** changed file types */
 
 interface BaseFileChange<Type extends string> extends Base<Type> {
-  hunks: Hunk[];
+  chunks: Chunk[];
 }
 
 export interface ChangedFile extends BaseFileChange<'ChangedFile'> {}
@@ -40,13 +40,13 @@ export type AnyFileChange = ChangedFile | AddedFile | DeletedFile | RenamedFile;
 
 /** hunk */
 
-export interface HunkPos {
+export interface ChunkPos {
   start: number;
   lines: number;
 }
 
-export interface Hunk extends Base<'Hunk'> {
-  addedPos: HunkPos;
-  deletedPos: HunkPos;
+export interface Chunk extends Base<'Hunk'> {
+  addedPos: ChunkPos;
+  deletedPos: ChunkPos;
   changes: AnyChange[];
 }`;
    const result = parseChunks(createContext(src));

    expect(result).not.toBe(null);
    expect(result).toMatchSnapshot();
  });

  it('should parse chunks 2.', () => {
    // prettier-ignore
    const src =
`@@ -18,7 +18,7 @@
 /** changed file types */
 
 interface BaseFileChange<Type extends string> extends Base<Type> {
-  hunks: Hunk[];
+  chunks: Chunk[];
 }
 
 export interface ChangedFile extends BaseFileChange<'ChangedFile'> {}
@@ -40,13 +40,13 @@ export type AnyFileChange = ChangedFile | AddedFile | DeletedFile | RenamedFile;
 
 /** hunk */
 
-export interface HunkPos {
+export interface ChunkPos {
   start: number;
   lines: number;
 }
 
-export interface Hunk extends Base<'Hunk'> {
-  addedPos: HunkPos;
-  deletedPos: HunkPos;
+export interface Chunk extends Base<'Hunk'> {
+  addedPos: ChunkPos;
+  deletedPos: ChunkPos;
   changes: AnyChange[];
 }`;
    const result = parseChunks(createContext(src));

    expect(result).not.toBe(null);
    expect(result).toMatchSnapshot();
  });
});
