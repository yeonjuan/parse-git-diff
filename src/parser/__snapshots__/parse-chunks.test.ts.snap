// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseChunks should parse chunks 2. 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "/** changed file types */",
        "lineAfter": 18,
        "lineBefore": 18,
        "type": "UnchangedLine",
      },
      Object {
        "content": "",
        "lineAfter": 19,
        "lineBefore": 19,
        "type": "UnchangedLine",
      },
      Object {
        "content": "interface BaseFileChange<Type extends string> extends Base<Type> {",
        "lineAfter": 20,
        "lineBefore": 20,
        "type": "UnchangedLine",
      },
      Object {
        "content": "  hunks: Hunk[];",
        "lineBefore": 21,
        "type": "DeletedLine",
      },
      Object {
        "content": "  chunks: Chunk[];",
        "lineAfter": 21,
        "type": "AddedLine",
      },
      Object {
        "content": "}",
        "lineAfter": 22,
        "lineBefore": 22,
        "type": "UnchangedLine",
      },
      Object {
        "content": "",
        "lineAfter": 23,
        "lineBefore": 23,
        "type": "UnchangedLine",
      },
      Object {
        "content": "export interface ChangedFile extends BaseFileChange<'ChangedFile'> {}",
        "lineAfter": 24,
        "lineBefore": 24,
        "type": "UnchangedLine",
      },
    ],
    "rangeAfter": Object {
      "lines": 7,
      "start": 18,
    },
    "rangeBefore": Object {
      "lines": 7,
      "start": 18,
    },
    "type": "Chunk",
  },
  Object {
    "changes": Array [
      Object {
        "content": "",
        "lineAfter": 40,
        "lineBefore": 40,
        "type": "UnchangedLine",
      },
      Object {
        "content": "/** hunk */",
        "lineAfter": 41,
        "lineBefore": 41,
        "type": "UnchangedLine",
      },
      Object {
        "content": "",
        "lineAfter": 42,
        "lineBefore": 42,
        "type": "UnchangedLine",
      },
      Object {
        "content": "export interface HunkPos {",
        "lineBefore": 43,
        "type": "DeletedLine",
      },
      Object {
        "content": "export interface ChunkPos {",
        "lineAfter": 43,
        "type": "AddedLine",
      },
      Object {
        "content": "  start: number;",
        "lineAfter": 44,
        "lineBefore": 44,
        "type": "UnchangedLine",
      },
      Object {
        "content": "  lines: number;",
        "lineAfter": 45,
        "lineBefore": 45,
        "type": "UnchangedLine",
      },
      Object {
        "content": "}",
        "lineAfter": 46,
        "lineBefore": 46,
        "type": "UnchangedLine",
      },
      Object {
        "content": "",
        "lineAfter": 47,
        "lineBefore": 47,
        "type": "UnchangedLine",
      },
      Object {
        "content": "export interface Hunk extends Base<'Hunk'> {",
        "lineBefore": 48,
        "type": "DeletedLine",
      },
      Object {
        "content": "  addedPos: HunkPos;",
        "lineBefore": 49,
        "type": "DeletedLine",
      },
      Object {
        "content": "  deletedPos: HunkPos;",
        "lineBefore": 50,
        "type": "DeletedLine",
      },
      Object {
        "content": "export interface Chunk extends Base<'Hunk'> {",
        "lineAfter": 48,
        "type": "AddedLine",
      },
      Object {
        "content": "  addedPos: ChunkPos;",
        "lineAfter": 49,
        "type": "AddedLine",
      },
      Object {
        "content": "  deletedPos: ChunkPos;",
        "lineAfter": 50,
        "type": "AddedLine",
      },
      Object {
        "content": "  changes: AnyChange[];",
        "lineAfter": 51,
        "lineBefore": 51,
        "type": "UnchangedLine",
      },
      Object {
        "content": "}",
        "lineAfter": 52,
        "lineBefore": 52,
        "type": "UnchangedLine",
      },
    ],
    "rangeAfter": Object {
      "lines": 13,
      "start": 40,
    },
    "rangeBefore": Object {
      "lines": 13,
      "start": 40,
    },
    "type": "Chunk",
  },
]
`;

exports[`parseChunks should parse chunks. 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "/** changed file types */",
        "lineAfter": 18,
        "lineBefore": 18,
        "type": "UnchangedLine",
      },
      Object {
        "content": "",
        "lineAfter": 19,
        "lineBefore": 19,
        "type": "UnchangedLine",
      },
      Object {
        "content": "interface BaseFileChange<Type extends string> extends Base<Type> {",
        "lineAfter": 20,
        "lineBefore": 20,
        "type": "UnchangedLine",
      },
      Object {
        "content": "  hunks: Hunk[];",
        "lineBefore": 21,
        "type": "DeletedLine",
      },
      Object {
        "content": "  chunks: Chunk[];",
        "lineAfter": 21,
        "type": "AddedLine",
      },
      Object {
        "content": "}",
        "lineAfter": 22,
        "lineBefore": 22,
        "type": "UnchangedLine",
      },
      Object {
        "content": "",
        "lineAfter": 23,
        "lineBefore": 23,
        "type": "UnchangedLine",
      },
      Object {
        "content": "export interface ChangedFile extends BaseFileChange<'ChangedFile'> {}",
        "lineAfter": 24,
        "lineBefore": 24,
        "type": "UnchangedLine",
      },
    ],
    "rangeAfter": Object {
      "lines": 7,
      "start": 18,
    },
    "rangeBefore": Object {
      "lines": 7,
      "start": 18,
    },
    "type": "Chunk",
  },
  Object {
    "changes": Array [
      Object {
        "content": "",
        "lineAfter": 40,
        "lineBefore": 40,
        "type": "UnchangedLine",
      },
      Object {
        "content": "/** hunk */",
        "lineAfter": 41,
        "lineBefore": 41,
        "type": "UnchangedLine",
      },
      Object {
        "content": "",
        "lineAfter": 42,
        "lineBefore": 42,
        "type": "UnchangedLine",
      },
      Object {
        "content": "export interface HunkPos {",
        "lineBefore": 43,
        "type": "DeletedLine",
      },
      Object {
        "content": "export interface ChunkPos {",
        "lineAfter": 43,
        "type": "AddedLine",
      },
      Object {
        "content": "  start: number;",
        "lineAfter": 44,
        "lineBefore": 44,
        "type": "UnchangedLine",
      },
      Object {
        "content": "  lines: number;",
        "lineAfter": 45,
        "lineBefore": 45,
        "type": "UnchangedLine",
      },
      Object {
        "content": "}",
        "lineAfter": 46,
        "lineBefore": 46,
        "type": "UnchangedLine",
      },
      Object {
        "content": "",
        "lineAfter": 47,
        "lineBefore": 47,
        "type": "UnchangedLine",
      },
      Object {
        "content": "export interface Hunk extends Base<'Hunk'> {",
        "lineBefore": 48,
        "type": "DeletedLine",
      },
      Object {
        "content": "  addedPos: HunkPos;",
        "lineBefore": 49,
        "type": "DeletedLine",
      },
      Object {
        "content": "  deletedPos: HunkPos;",
        "lineBefore": 50,
        "type": "DeletedLine",
      },
      Object {
        "content": "export interface Chunk extends Base<'Hunk'> {",
        "lineAfter": 48,
        "type": "AddedLine",
      },
      Object {
        "content": "  addedPos: ChunkPos;",
        "lineAfter": 49,
        "type": "AddedLine",
      },
      Object {
        "content": "  deletedPos: ChunkPos;",
        "lineAfter": 50,
        "type": "AddedLine",
      },
      Object {
        "content": "  changes: AnyChange[];",
        "lineAfter": 51,
        "lineBefore": 51,
        "type": "UnchangedLine",
      },
      Object {
        "content": "}",
        "lineAfter": 52,
        "lineBefore": 52,
        "type": "UnchangedLine",
      },
    ],
    "rangeAfter": Object {
      "lines": 13,
      "start": 40,
    },
    "rangeBefore": Object {
      "lines": 13,
      "start": 40,
    },
    "type": "Chunk",
  },
]
`;
